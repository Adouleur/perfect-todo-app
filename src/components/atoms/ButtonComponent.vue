<script setup>
import { defineProps } from 'vue';

const props = defineProps({
  class: { type: String, default: '' }
});
</script>

<template>
  <button type="button" :class="props.class">
    <slot></slot>
  </button>
</template>

<style scoped>
button {
  padding: 5px 10px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}
</style>
