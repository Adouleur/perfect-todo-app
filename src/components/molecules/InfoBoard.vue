<script setup>
import { defineProps } from 'vue';

const props = defineProps({
  tasksLeft: { type: Number, required: true },
  setFilter: { type: Function, required: true },
  currentFilter: { type: String, required: true }
});
</script>

<template>
  <div class="info-board">
    <div class="filters">
      <button :class="{ active: currentFilter === 'all' }" @click="setFilter('all')">All</button>
      <button :class="{ active: currentFilter === 'active' }" @click="setFilter('active')">
        Active
      </button>
      <button :class="{ active: currentFilter === 'completed' }" @click="setFilter('completed')">
        Completed
      </button>
    </div>
    <div class="tasks-left">{{ tasksLeft }} {{ tasksLeft === 1 ? 'task' : 'tasks' }} left</div>
  </div>
</template>

<style scoped>
.info-board {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 16px;
  background-color: #f8f8f8;
  border-top: 1px solid #ddd;
  border-radius: 0 0 8px 8px;
}

.filters {
  display: flex;
  gap: 10px;
}

button {
  padding: 5px 10px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

button.active {
  background: #007bff;
  color: white;
}

.tasks-left {
  font-size: 14px;
  color: #555;
}
</style>
