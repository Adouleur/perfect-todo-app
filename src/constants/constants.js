export const filters = ['all', 'active', 'completed'];
