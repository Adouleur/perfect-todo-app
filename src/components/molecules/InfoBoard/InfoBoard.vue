<script setup>
import { defineProps } from 'vue';
import { useStore } from 'vuex';
import ButtonComponent from '@/components/atoms/ButtonComponent/ButtonComponent.vue';
import { filters } from '@/constants/constants';
import { capitalizeFirstLetter } from '@/helpers/capitalizeFirstLetter';

const props = defineProps({
  tasksLeft: { type: Number, required: true },
  setFilter: { type: Function, required: true },
  currentFilter: { type: String, required: true }
});

const store = useStore();
const clearCompleted = () => store.dispatch('clearCompletedTodos');
</script>

<template>
  <div class="info-board">
    <div class="filters">
      <ButtonComponent
        v-for="filter in filters"
        :key="filter"
        :class="{ active: currentFilter === filter }"
        @click="setFilter(filter)"
        :data-cy="`filter-${filter}`"
      >
        {{ capitalizeFirstLetter(filter) }}
      </ButtonComponent>

      <ButtonComponent @click="clearCompleted" data-cy="clear-completed"
        >Clear Completed</ButtonComponent
      >
    </div>
    <div class="tasks-left">{{ tasksLeft }} {{ tasksLeft === 1 ? 'task' : 'tasks' }} left</div>
  </div>
</template>

<style scoped>
.info-board {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 16px;
  background-color: #f8f8f8;
  border-top: 1px solid #ddd;
  border-radius: 0 0 8px 8px;
}

.filters {
  display: flex;
  gap: 10px;
}

button.active {
  background: #007bff;
  color: white;
}

.tasks-left {
  font-size: 14px;
  color: #555;
}
</style>
