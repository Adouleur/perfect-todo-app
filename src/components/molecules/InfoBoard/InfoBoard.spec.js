import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { describe, it, expect, beforeEach, vi } from 'vitest';
import ButtonComponent from '@/components/atoms/ButtonComponent/ButtonComponent.vue';
import { filters } from '@/constants/constants';
import InfoBoard from '@/components/molecules/InfoBoard/InfoBoard.vue';

describe('InfoBoard.vue', () => {
  let actions;
  let mutations;
  let store;

  beforeEach(() => {
    // Mock the Vuex store
    actions = {
      clearCompletedTodos: vi.fn()
    };
    mutations = {
      setFilter: vi.fn()
    };
    store = createStore({
      state: {
        filter: filters[0] // Default filter
      },
      actions,
      mutations,
      getters: {
        incompleteTodos: () => 3 // Assume 3 incomplete tasks
      }
    });
  });

  it('renders the correct number of tasks left', () => {
    const wrapper = mount(InfoBoard, {
      props: {
        tasksLeft: 3,
        setFilter: vi.fn(),
        currentFilter: filters[0]
      },
      global: {
        plugins: [store]
      }
    });

    // Check if the tasks left count is rendered correctly
    expect(wrapper.find('.tasks-left').text()).toBe('3 tasks left');
  });

  it('calls the setFilter method when a filter button is clicked', async () => {
    const setFilter = vi.fn();
    const wrapper = mount(InfoBoard, {
      props: {
        tasksLeft: 3,
        setFilter,
        currentFilter: filters[0]
      },
      global: {
        plugins: [store]
      }
    });

    const button = wrapper.findAllComponents(ButtonComponent).at(0); // First filter button
    await button.trigger('click');

    // Check if the setFilter function is called with the correct argument
    expect(setFilter).toHaveBeenCalledWith(filters[0]);
  });

  it('dispatches the clearCompletedTodos action when "Clear Completed" button is clicked', async () => {
    const wrapper = mount(InfoBoard, {
      props: {
        tasksLeft: 3,
        setFilter: vi.fn(),
        currentFilter: filters[0]
      },
      global: {
        plugins: [store]
      }
    });

    const clearCompletedButton = wrapper.find('[data-cy="clear-completed"]');
    await clearCompletedButton.trigger('click');

    // Check if the store action is dispatched correctly
    expect(actions.clearCompletedTodos).toHaveBeenCalled();
  });

  it('renders the correct filter buttons and applies the active class', () => {
    const wrapper = mount(InfoBoard, {
      props: {
        tasksLeft: 3,
        setFilter: vi.fn(),
        currentFilter: filters[1] // Assuming the active filter is the second one
      },
      global: {
        plugins: [store]
      }
    });

    const buttons = wrapper.findAllComponents(ButtonComponent);
    filters.forEach((filter, index) => {
      expect(buttons.at(index).text()).toBe(filter.charAt(0).toUpperCase() + filter.slice(1));
      if (filter === filters[1]) {
        expect(buttons.at(index).classes()).toContain('active');
      }
    });
  });
});
