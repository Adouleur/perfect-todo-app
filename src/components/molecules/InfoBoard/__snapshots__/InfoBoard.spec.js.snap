// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfoBoard.vue > matches the snapshot 1`] = `
"<div data-v-41083afe="" class="info-board">
  <div data-v-41083afe="" class="filters"><button data-v-79b848db="" data-v-41083afe="" type="button" class="active" data-cy="filter-all">All</button><button data-v-79b848db="" data-v-41083afe="" type="button" class="" data-cy="filter-active">Active</button><button data-v-79b848db="" data-v-41083afe="" type="button" class="" data-cy="filter-completed">Completed</button><button data-v-79b848db="" data-v-41083afe="" type="button" class="" data-cy="clear-completed">Clear Completed</button></div>
  <div data-v-41083afe="" class="tasks-left">3 tasks left</div>
</div>"
`;
